SymCryptProvider/src/kdf/p_scossl_hkdf.c

```diff
@@ -341,6 +330,367 @@ SCOSSL_STATUS p_scossl_hkdf_set_ctx_params(_Inout_ SCOSSL_PROV_HKDF_CTX *ctx, co
+ * } HkdfLabel;
+ *
+ */
+static
+SCOSSL_STATUS p_scossl_tls13_hkdf_expand(_In_ SCOSSL_HKDF_CTX *ctx, 
+                                         _Out_writes_bytes_(keylen) unsigned char *key, size_t keylen)
+{
+    SYMCRYPT_ERROR scError = SYMCRYPT_NO_ERROR;
+    PCSYMCRYPT_MAC symcryptHmacAlg = NULL;
+    SYMCRYPT_HKDF_EXPANDED_KEY scExpandedKey;
+    SIZE_T labelLen = 0;
+    SIZE_T totalLen = 0;
+        
+    BYTE hkdflabel[HKDF_MAXBUF];
+    SIZE_T hkdflabellen = 0;
+    
+    if (ctx->md == NULL) 
+    {
+        SCOSSL_LOG_ERROR(SCOSSL_ERR_F_HKDF_DERIVE, ERR_R_INTERNAL_ERROR, "Missing Digest");
+        return SCOSSL_FAILURE;
+    }
+
+    if (ctx->pbKey == NULL) 
+    {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return SCOSSL_FAILURE;
+    }
+
+    symcryptHmacAlg = scossl_get_symcrypt_hmac_algorithm(EVP_MD_type(ctx->md));
+    if (symcryptHmacAlg == NULL) 
+    {
+        return SCOSSL_FAILURE;
+    }
+
+    labelLen = ctx->cbPrefix + ctx->cbLabel;
+    
+    // Ensure this value does not exceed 0xFF, as only the least-significant byte is copied into hkdflabel.
+    // If the value exceeds 0xFF, it will overflow and corrupt the label encoding.
+    if (labelLen > 0xFF)
+    {
+        SCOSSL_LOG_ERROR(SCOSSL_ERR_F_HKDF_DERIVE, ERR_R_INTERNAL_ERROR, "Label+Prefix length exceeds limit");
+        return SCOSSL_FAILURE;
+    }
+    
+    // 2 bytes for output length, 1 byte for label length, and 1 byte for context length
+    totalLen = 2 + 1 + labelLen + 1 + ctx->cbData;
+    if (totalLen > HKDF_MAXBUF) 
+    {
+        SCOSSL_LOG_ERROR(SCOSSL_ERR_F_HKDF_DERIVE, ERR_R_INTERNAL_ERROR,
+                         "Total size exceeds maximum buffer size allowed");
+        return SCOSSL_FAILURE;
+    }
+
+    // Output length (2 bytes) in big endian format
+    SYMCRYPT_STORE_MSBFIRST16(&hkdflabel[hkdflabellen], keylen);
+    hkdflabellen += 2;
+
+    // Label length
+    hkdflabel[hkdflabellen++] = (BYTE)labelLen;
+
+    // Label = prefix + label
+    memcpy(hkdflabel + hkdflabellen, ctx->pbPrefix, ctx->cbPrefix);
+    hkdflabellen += ctx->cbPrefix;
+    memcpy(hkdflabel + hkdflabellen, ctx->pbLabel, ctx->cbLabel);
+    hkdflabellen += ctx->cbLabel;
+    
+    if (ctx->cbData > 0xFF)
+    {
+        SCOSSL_LOG_ERROR(SCOSSL_ERR_F_HKDF_DERIVE, ERR_R_INTERNAL_ERROR, "context length exceeds limit");
+        return SCOSSL_FAILURE;
+    }
+    // Context length
+    hkdflabel[hkdflabellen++] = (BYTE)ctx->cbData;
+
+    // Context
+    if (ctx->cbData > 0) 
+    {
+        memcpy(hkdflabel + hkdflabellen, ctx->pbData, ctx->cbData);
+        hkdflabellen += ctx->cbData;
+    }
+
+    // Expand PRK
+    scError = SymCryptHkdfPrkExpandKey(
+        &scExpandedKey,
+        symcryptHmacAlg,
+        ctx->pbKey, ctx->cbKey);
+    if (scError != SYMCRYPT_NO_ERROR)
+    {
+        SCOSSL_LOG_ERROR(SCOSSL_ERR_F_HKDF_DERIVE, ERR_R_INTERNAL_ERROR, "SymCrypt expand failed");
```

Comment:
>SCOSSL_LOG_ERROR(SCOSSL_ERR_F_HKDF_DERIVE, ERR_R_INTERNAL_ERROR, "SymCrypt expand failed"); [](http://example.com/codeflow?start=8&length=91)

Use `SCOSSL_PROV_LOG_SYMCRYPT_ERROR` for SymCrypt failures

```suggestion
SCOSSL_PROV_LOG_SYMCRYPT_ERROR("SymCryptHkdfPrkExpandKey failed, scError);
```